src/sources/kafka.rs

```diff
@@ -237,34 +242,59 @@ async fn kafka_source(
     acknowledgements: bool,
 ) -> Result<(), ()> {
     let consumer = Arc::new(consumer);
-    let (finalizer, mut ack_stream) =
-        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, shutdown.clone());
-    let finalizer = finalizer.map(Arc::new);
-    if let Some(finalizer) = &finalizer {
+    let (mut finalizer, mut ack_stream) =
+        OrderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, None);
+
+    let (tx_callbacks, mut rx_callbacks) = mpsc::unbounded_channel();
+
+    if let Some(_) = &finalizer {
         consumer
             .context()
-            .finalizer
-            .set(Arc::clone(finalizer))
-            .expect("Finalizer is only set once");
+            .callback_channel
+            .set(tx_callbacks)
+            .expect("callback_channel is only set once");
     }
 
     let mut stream = consumer.stream();
     let keys = Keys::from(log_schema(), &config);
 
+    let mut shutting_down = false;
     loop {
         tokio::select! {
-            _ = &mut shutdown => break,
-            entry = ack_stream.next() => if let Some((status, entry)) = entry {
-                if status == BatchStatus::Delivered {
-                    if let Err(error) =
-                        consumer.store_offset(&entry.topic, entry.partition, entry.offset)
-                    {
-                        emit!(KafkaOffsetUpdateError { error });
+            _ = &mut shutdown => {
```

Comment:
This needs a guard clause, so that it doesn't run again after `shutting_down` is `true`.